src/main/java/io/cdap/plugin/scd2/SCD2Iterator.java

```diff
@@ -60,22 +62,64 @@ protected StructuredRecord computeNext() {
     cur = next != null ? next : records.next();
     next = records.hasNext() ? records.next() : null;
 
+    deduplicateCurRecord();
+    deduplicateNextRecord();
+
+    // if key changes, clean up the table to free memory
+    if (prev != null && !prev._1().equals(cur._1())) {
+      valTable.row(prev._1()).clear();
+    }
+
+    return computeRecord(cur._1(),
+                         prev != null && prev._1().equals(cur._1()) ? prev._2() : null,
+                         cur._2(),
+                         next != null && next._1().equals(cur._1()) ? next._2() : null);
+  }
+
+  /**
+   * Find the cur record that will not get deduplicated. This record will be used to returned and the end time
+   * will be computed for it.
+   *
+   * For example:
+   * 1. id: 1, name: john, isTarget: false (should get dedup)
+   * 2. id: 1, name: john, isTarget: true (should be preserved)
+   * 3. id: 1, name: john, isTarget: false (should get dedup)
+   * 4. id: 1, name: sam, isTarget: false (should get dedup)
+   * 5. id: 1, name: sam, isTarget: false (should be preserved)
+   *
+   * After this method, the prev, cur, next will be pointed to:
+   * 1.(prev) id: 1, name: john, isTarget: false (should get dedup)
+   * 2.(cur) id: 1, name: john, isTarget: true (should be preserved and returned)
+   * 3.(next) id: 1, name: john, isTarget: false (should get dedup)
```

Comment:
I think it's better if the contract is that the iterator advances to 4 after the method is called and not 3. Then the code becomes something like:

```
private Tuple2<SCD2Key, StructuredRecord> dedupeNext() {
  Tuple2<SCD2Key, StructuredRecord> currentElement = records.next();

  while (conf.deduplicate() && records.hasNext()) {
    Tuple2<SCD2Key, StructuredRecord> nextElement = records.next();
    if (isTarget(nextElement)) {
      currentElement = nextElement;
      continue;
    }
    if (isDiff(currentElement, nextElement) {
      break;
    }
  }

  return currentElement;
}

```

Then you just need need the other method with it's prev and next tracking, and only need to call this once per computeNext() call (except for the very first time, when you need to call it once to initialize cur and once to initialize next).

```
protected StructuredRecord computeNext() {
  ...
  prev = cur;
  if (prev = null) {
    cur = dedupeNext();
  } else {
    cur = next;
  }
  next = records.hasNext() ? dedupeNext() : null;
}

```